dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ThreadCreation.java

```diff
@@ -98,156 +103,121 @@ public void run(Program program) {
 
                 final Label joinCase = EventFactory.newLabel("__joinWithT" + tid + "#" + joinCounter);
-
-                // Construct a spinloop waiting for the joining thread.
-                // FIXME: It serves as a temporary solution by constructing the spinloop as how it should be
-                //        after the unrolling pass so that the processing pipeline does not have to be changed.
-                final Event loopBound = EventFactory.Svcomp.newLoopBound(expressions.makeValue(1, types.getArchType()));
-                final String loopLabelPrefix = "l_waitForT" + tid + "#" + joinCounter;
-                final Label waitingLoopBegin = EventFactory.newLabel(String.format("%s%s%s%s%s", loopLabelPrefix,
-                        LOOP_LABEL_IDENTIFIER, LOOP_INFO_SEPARATOR, LOOP_INFO_ITERATION_SUFFIX, 1));
-                waitingLoopBegin.addTags(Tag.NOOPT);
-                final Label joinFailedMarker = EventFactory.newLabel(loopLabelPrefix + ".failed");
-                final Event terminator = EventFactory.newGoto((Label) thread.getExit());
-                terminator.addTags(Tag.SPINLOOP, Tag.NONTERMINATION);
-                final Label endMarker = EventFactory.newLabel(String.format("%s%s%s%s", loopLabelPrefix,
-                        LOOP_LABEL_IDENTIFIER, LOOP_INFO_SEPARATOR, LOOP_INFO_BOUND_SUFFIX));
-                endMarker.addTags(Tag.NOOPT);
-                final Event boundEvent = newJump(expressions.makeTrue(), (Label) thread.getExit());
-                boundEvent.addTags(Tag.BOUND, Tag.NONTERMINATION, Tag.NOOPT);
-
-                final List<Event> caseBody = eventSequence(
-                        joinCase,
-                        loopBound,
-                        waitingLoopBegin,
-                        newAcquireLoad(joinDummyReg, comAddrOfThreadToJoinWith),
-                        newJump(joinDummyReg, joinFailedMarker),
-                        EventFactory.newGoto(joinEnd),
-                        joinFailedMarker,
-                        terminator,
-                        endMarker,
-                        boundEvent
-                );
-                tid2joinCases.put(tidCandidate, caseBody);
+                final List<Event> caseBody;
+                if (!data.thread().getFunctionType().getReturnType().equals(retValType)) {
+                    // Wrong return type
+                    caseBody = eventSequence(
+                            joinCase,
+                            newLocal(statusRegister, invalidRetType),
+                            EventFactory.newGoto(joinEnd)
+                    );
+                } else {
+                    // Successful join
+                    caseBody = eventSequence(
+                            joinCase,
+                            newThreadJoin(retValRegister, data.thread()),
+                            newAcquireLoad(syncRegister, data.comAddress),
+                            newAssume(expressions.makeNot(syncRegister)),
```

Comment:
I don't understand why we need this assume

Reply:
To establish synchronization without a spinloop. `ThreadJoin` will block until the target thread terminates, but it will not give us synchronization edges (e.g. `rf/hb`).
Notice that this is basically the code we had before we used a spinloop, but back then the code tried to block & synchronize, and now the two parts are done by separate events.